Add support for the `allow-downloads` iframe sandbox attribute

robert-jenner opened the pull request:
#### 58e9d98d9ccae4fdfae02d8d51e7ecd2e273f5c5
<pre>
Add support for the `allow-downloads` iframe sandbox attribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=259155">https://bugs.webkit.org/show_bug.cgi?id=259155</a>
rdar://109414664

Reviewed by Brent Fulgham.

Add support for the `allow-downloads` iframe sandbox attribute so that downloads
are now forbidden by default in sandboxed iframes and this attribute is required
to allow them. This aligns our behavior with Chrome and Firefox.

* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-expected.txt: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-in-popup-expected.txt: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-in-popup-noopener-expected.txt: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-in-popup-noopener.html: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-in-popup.html: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed.html: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-not-allowed-expected.txt: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-not-allowed.html: Added.
Add layout test coverage.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_anchor_download_block_downloads.tentative-expected.txt:
Rebaseline due to more PASS lines. I investigated the remaining failure and confirmed that no download is actually occurring for any of the subtests, as expected.
I see our code cancelling the NetworkDataTask. However, the test doesn&apos;t seem to detect the stream closing. It may be an issue as CFNetwork level.

* Source/WebCore/dom/SecurityContext.cpp:
(WebCore::SecurityContext::isSupportedSandboxPolicy):
(WebCore::SecurityContext::parseSandboxPolicy):
* Source/WebCore/dom/SecurityContext.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::continueAfterContentPolicy):
* Source/WebCore/loader/PolicyChecker.cpp:
(WebCore::FrameLoader::PolicyChecker::checkNavigationPolicy):
(WebCore::FrameLoader::PolicyChecker::checkNewWindowPolicy):

* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::decidePolicyForNavigationResponse):
Update WebKitTestRunner to trigger a download when a resource response contains
`Content-Disposition: attachment`, similarly to what a browser would do and what
WPT tests expect. By default, we would only trigger a download if the mime type
couldn&apos;t be displayed. However, iframe_sandbox_anchor_download_block_downloads.tentative.html
WPT test uses both a `Content-Disposition: attachment` HTTP header and a displayable
MIME type and rightly expects a download.

Originally-landed-as: 265870.84@safari-7616-branch (9d46b7777bca). rdar://116425379
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/58e9d98d9ccae4fdfae02d8d51e7ecd2e273f5c5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22129 "Passed style check") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22337 "Hash 58e9d98d for PR 18916 does not build (failure)") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23198 "Hash 58e9d98d for PR 18916 does not build (failure)") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24021 "Hash 58e9d98d for PR 18916 does not build (failure)") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20486 "Hash 58e9d98d for PR 18916 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22374 "Passed tests") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26588 "Hash 58e9d98d for PR 18916 does not build (failure)") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22654 "Hash 58e9d98d for PR 18916 does not build (failure)") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/24021 "Hash 58e9d98d for PR 18916 does not build (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22359 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/23/builds/26588 "Hash 58e9d98d for PR 18916 does not build (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/14/builds/23198 "Hash 58e9d98d for PR 18916 does not build (failure)") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24871 "Hash 58e9d98d for PR 18916 does not build (failure)") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/23/builds/26588 "Hash 58e9d98d for PR 18916 does not build (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/14/builds/23198 "Hash 58e9d98d for PR 18916 does not build (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/24871 "Hash 58e9d98d for PR 18916 does not build (failure)") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/23/builds/26588 "Hash 58e9d98d for PR 18916 does not build (failure)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/14/builds/23198 "Hash 58e9d98d for PR 18916 does not build (failure)") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/24871 "Hash 58e9d98d for PR 18916 does not build (failure)") | 
| | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20712 "Hash 58e9d98d for PR 18916 does not build (failure)") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/16/builds/22654 "Hash 58e9d98d for PR 18916 does not build (failure)") | | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20066 "Hash 58e9d98d for PR 18916 does not build (failure)") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/14/builds/23198 "Hash 58e9d98d for PR 18916 does not build (failure)") | | 
| | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24274 "Hash 58e9d98d for PR 18916 does not build (failure)") | | | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20660 "Hash 58e9d98d for PR 18916 does not build (failure)") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
```diff
@@ -42,7 +42,7 @@ interface TestRunner {
     undefined waitUntilDownloadFinished();
     undefined setShouldLogDownloadCallbacks(boolean value);
     undefined setShouldLogDownloadSize(boolean value);
-    undefined setShouldLogDownloadExpectedSize(boolean value);
```
You need to introduce back this line here.

cdumez commented on Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
```diff
@@ -233,7 +233,7 @@ class TestRunner : public JSWrappable {
     bool shouldFinishAfterDownload() const { return m_shouldFinishAfterDownload; }
     void setShouldLogDownloadCallbacks(bool);
     void setShouldLogDownloadSize(bool);
-    void setShouldLogDownloadExpectedSize(bool);
```
This needs to be re-introduced.

cdumez commented on Tools/WebKitTestRunner/TestController.h:
```diff
@@ -201,7 +201,7 @@ class TestController {
     void setShouldLogCanAuthenticateAgainstProtectionSpace(bool shouldLog) { m_shouldLogCanAuthenticateAgainstProtectionSpace = shouldLog; }
     void setShouldLogDownloadCallbacks(bool shouldLog) { m_shouldLogDownloadCallbacks = shouldLog; }
     void setShouldLogDownloadSize(bool shouldLog) { m_shouldLogDownloadSize = shouldLog; }
-    void setShouldLogDownloadExpectedSize(bool shouldLog) { m_shouldLogDownloadExpectedSize = shouldLog; }
```
This needs to be re-introduced.

cdumez committed:
Add support for the `allow-downloads` iframe sandbox attribute
https://bugs.webkit.org/show_bug.cgi?id=259155
rdar://109414664

Reviewed by Brent Fulgham.

Add support for the `allow-downloads` iframe sandbox attribute so that downloads
are now forbidden by default in sandboxed iframes and this attribute is required
to allow them. This aligns our behavior with Chrome and Firefox.

* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-expected.txt: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-in-popup-expected.txt: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-in-popup-noopener-expected.txt: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-in-popup-noopener.html: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-in-popup.html: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed.html: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-not-allowed-expected.txt: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-not-allowed.html: Added.
Add layout test coverage.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_anchor_download_block_downloads.tentative-expected.txt:
Rebaseline due to more PASS lines. I investigated the remaining failure and confirmed that no download is actually occurring for any of the subtests, as expected.
I see our code cancelling the NetworkDataTask. However, the test doesn't seem to detect the stream closing. It may be an issue as CFNetwork level.

* Source/WebCore/dom/SecurityContext.cpp:
(WebCore::SecurityContext::isSupportedSandboxPolicy):
(WebCore::SecurityContext::parseSandboxPolicy):
* Source/WebCore/dom/SecurityContext.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::continueAfterContentPolicy):
* Source/WebCore/loader/PolicyChecker.cpp:
(WebCore::FrameLoader::PolicyChecker::checkNavigationPolicy):
(WebCore::FrameLoader::PolicyChecker::checkNewWindowPolicy):

* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::decidePolicyForNavigationResponse):
Update WebKitTestRunner to trigger a download when a resource response contains
`Content-Disposition: attachment`, similarly to what a browser would do and what
WPT tests expect. By default, we would only trigger a download if the mime type
couldn't be displayed. However, iframe_sandbox_anchor_download_block_downloads.tentative.html
WPT test uses both a `Content-Disposition: attachment` HTTP header and a displayable
MIME type and rightly expects a download.

Originally-landed-as: 265870.84@safari-7616-branch (9d46b7777bca). rdar://116425379
